pkg/executorconfig/v1alpha1/executor_config.go

```diff
@@ -0,0 +1,169 @@
+// of ObjectName, ObjectNamespace, ObjectKind passed to cast at runtime.
+type Resource struct {
+	// ObjectName is name of the resource which is going to be updated
+	// use - { .TaskConfig.ObjectName } to get this value in runtask
+	ObjectName string `yaml:"objectName"`
+	// ObjectNamespace is namespace of the resource which is going to be updated
+	// use - { .TaskConfig.ObjectNamespace } to get this value in runtask
+	ObjectNamespace string `yaml:"objectNamespace"`
+	// ObjectKind is kind of the resource which is going to be updated
+	// use - { .TaskConfig.ObjectKind } to get this value in runtask
+	ObjectKind string `yaml:"objectKind"`
+	// CastemplateName is cast which will be used to update this resource
+	CastemplateName string `yaml:"castemplate"`
+}
+
+// ExecutorConfig provides configuration to all the castemplate for updating mentioned
+// resources values of BaseVersion, TargetVersion, UpgradeTaskCRName, RetryOnError
+// RollbackOnError passed to cast at runtime.
+type ExecutorConfig struct {
+	// BaseVersion is current version of the resource
+	// use - { .TaskConfig.BaseVersion } to get this value in runtask
+	BaseVersion string `yaml:"baseVersion"`
+	// TargetVersion is the version in which resource is going to be updated
+	// use - { .TaskConfig.TargetVersion } to get this value in runtask
+	TargetVersion string `yaml:"targetVersion"`
+	// UpgradeTaskCRName is result output cr's name in which o/p of each task will be stored
+	// use - { .TaskConfig.UpgradeTaskCRName } to get this value in runtask
+	UpgradeTaskCRName string `yaml:"upgradeTaskName"`
+	// RetryOnError use to take decision on what to do if an error occurred during update
+	// use - { .TaskConfig.RetryOnError } to get this value in runtask
+	RetryOnError bool `yaml:"retryOnError"`
+	// RollbackOnError use to take decision on what to do if an error occurred during update
+	// use - { .TaskConfig.RollbackOnError } to get this value in runtask
+	RollbackOnError bool `yaml:"rollbackOnError"`
+	// OpenebsConfig is openebs specific configuration which is directly passed to castemplate
+	// use - { .Config.<key(name)>.value } to get this value in runtask
+	OpenebsConfig []apis.Config `yaml:"openebsConfig"`
+	// Resources contains list of resources which are going to be updated
+	Resources []Resource `yaml:"resources"`
+}
+
+// NewConfigFromRaw returns ExecutorConfig structure by Unmarshaling raw data. It can be as yaml or json format
+func NewConfigFromRaw(data []byte) (ec *ExecutorConfig, err error) {
+	ec = &ExecutorConfig{}
+	if json.Valid(data) {
+		err = json.Unmarshal(data, ec)
+	} else {
+		err = yaml.Unmarshal(data, ec)
+	}
+	return
+}
+
+// NewConfig returns empty struct of ExecutorConfig type
+func NewConfig() *ExecutorConfig {
+	ec := &ExecutorConfig{}
+	return ec
+}
+
+// WithBaseVersion adds baseVersion in ExecutorConfig object and returns updated ExecutorConfig object
+func (ec *ExecutorConfig) WithBaseVersion(baseVersion string) *ExecutorConfig {
```

Comment:
Use these as methods of builder instance

```go
// pkg/upgrade/v1alpha1

import (
  apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
)

type upgrade struct {
  object *apis.Upgrade
}

type buildOption func(*upgrade)

func WithUpgradeObject(api *apis.Upgrade) buildOpt {
  return func(u *upgrade) {
    u.object = api
  }
}

func New(opts ...buildOpt) *upgrade {
  u := &upgrade{}
  for _, o := range opts {
    u(o)
  }
  return u
}

type builder struct {
  upgrade *upgrade
}

func Builder() *builder {
  return &builder{upgrade: &apis.Upgrade{}}
}

func (b *builder) WithBaseVersion(baseVer string) *builder {
  // ...
}
```

Reply:
Caller code need to decide if New() should be invoked versus Builder().
Mostly when a configmap's data is unmarshalled, then entire api object is populated. So Builder is not required in this case.

Reply:
will do in next pr.